src/rust/engine/rule_graph/src/rules.rs

```diff
@@ -22,26 +23,61 @@ pub trait TypeId:
     I: Iterator<Item = Self>;
 }
 
-pub trait DependencyKey:
-  Clone + Copy + Debug + Display + Hash + Ord + Eq + Sized + 'static
-{
-  type TypeId: TypeId;
+#[derive(DeepSizeOf, Eq, Hash, PartialEq, Clone, Debug, PartialOrd, Ord)]
+pub struct DependencyKey<T: TypeId> {
+  product: T,
+  provided_params: SmallVec<[T; 2]>,
+}
 
-  ///
-  /// Generate a DependencyKey for a dependency at the "root" of the RuleGraph, which represents an
-  /// entrypoint into the set of installed Rules.
-  ///
-  fn new_root(product: Self::TypeId) -> Self;
+impl<T: TypeId> DependencyKey<T> {
+  pub fn new(product: T) -> Self {
+    DependencyKey {
+      product,
+      provided_params: SmallVec::default(),
+    }
+  }
+
+  pub fn new_with_params<I: IntoIterator<Item = T>>(product: T, provided_params: I) -> Self {
+    let mut provided_params = provided_params.into_iter().collect::<SmallVec<[T; 2]>>();
+    provided_params.sort();
+
+    if cfg!(debug_assertions) {
+      let original_len = provided_params.len();
+      provided_params.dedup();
+      if original_len != provided_params.len() {
+        panic!("Expected unique provided params.");
+      }
+    }
+
+    DependencyKey {
+      product,
+      provided_params,
+    }
+  }
 
   ///
   /// Returns the product (output) type for this dependency.
   ///
-  fn product(&self) -> Self::TypeId;
+  pub fn product(&self) -> T {
+    self.product
+  }
 
   ///
   /// Returns the Param (input) type for this dependency, if it provides one.
   ///
-  fn provided_param(&self) -> Option<Self::TypeId>;
+  pub fn provided_param(&self) -> Option<T> {
+    self.provided_params.get(0).cloned()
```

Comment:
What is the semantic difference between index 0 and other indices in `self.provided_params`? If there are multiple entries in `provided_params`, do callers of this method need to account for that?

Reply:
This is to bridge until #7490 and #12946 are actually implemented. As it stands, all callers have either zero or one provided param.

Reply:
Then worth a comment?